be/src/vec/exprs/vcompound_pred.h

```diff
@@ -53,6 +53,75 @@ class VCompoundPred : public VectorizedFnCall {
+    //   2. when meet 'and' conjunct, function with column b can not apply inverted index
+    //      eg. a and hash(b)=1, if b can apply index, but hash(b)=1 is not for index, so b should not be extracted
+    //          but a and array_contains(b, 1), b can be applied inverted index, which b can be extracted
+    Status eval_inverted_index(
+            VExprContext* context,
+            const std::unordered_map<ColumnId, std::pair<vectorized::NameAndTypePair,
+                                                         segment_v2::InvertedIndexIterator*>>&
+                    colId_to_inverted_index_iter,
+            uint32_t num_rows, roaring::Roaring* bitmap) const override {
+        if (_op == TExprOpcode::COMPOUND_OR) {
+            for (auto child : _children) {
+                std::shared_ptr<roaring::Roaring> child_roaring =
+                        std::make_shared<roaring::Roaring>();
+                Status st = child->eval_inverted_index(context, colId_to_inverted_index_iter,
+                                                       num_rows, child_roaring.get());
+                if (!st.ok()) {
+                    continue;
+                }
+                *bitmap |= *child_roaring;
+                if (!child_roaring->isEmpty()) {
+                    // means inverted index filter do not reduce any rows
+                    // the left expr no need to be extracted by inverted index,
+                    // and cur roaring is all rows which means this inverted index is not useful,
+                    // do not need to calculate with res bitmap
+                    return Status::OK();
+                }
+            }
+        } else if (_op == TExprOpcode::COMPOUND_AND) {
+            for (auto child : _children) {
+                std::shared_ptr<roaring::Roaring> child_roaring =
+                        std::make_shared<roaring::Roaring>();
+                Status st = child->eval_inverted_index(context, colId_to_inverted_index_iter,
+                                                       num_rows, child_roaring.get());
+                if (!st.ok()) {
+                    continue;
+                }
+                *bitmap &= *child_roaring;
+                if (child_roaring->isEmpty()) {
+                    // the left expr no need to be extracted by inverted index, just return 0 rows
+                    // res bitmap will be zero
+                    return Status::OK();
+                }
+            }
+        } else if (_op == TExprOpcode::COMPOUND_NOT) {
+            std::shared_ptr<roaring::Roaring> child_roaring = std::make_shared<roaring::Roaring>();
+            Status st = _children[0]->eval_inverted_index(context, colId_to_inverted_index_iter,
+                                                          num_rows, child_roaring.get());
+            if (!st.ok()) {
+                return st;
+            }
+            *bitmap -= *child_roaring;
```

Comment:
-= child_roaring or -child_roaring ?

Reply:
-= is override function : which will call  api::roaring_bitmap_andnot_inplace(&roaring, &r.roaring);